File: js/src/select/recursion.js
import { keyGroups } from "../query/groups.js";
import { isWellFormedBranch } from "../schema.js";
import { matchesRegex } from "../match.js";

/**
 * One `~` traversal: a closure over the tablet that stores the
 * relationship between the query base and the named field.
 * Mirrors the Rust `Traversal` in rs/src/dataset/select/recursion.rs.
 */
export class Traversal {
  constructor(stops, neighbors) {
    // Stop patterns (regex strings): a value matching any of these
    // is neither returned nor expanded.
    this.stops = stops;

    // Two-way adjacency over the tablet lines. The direction in
    // which the tablet stores the relationship does not affect
    // the result.
    this.neighbors = neighbors;

    // Values this traversal has already expanded.
    this.visited = new Set();
  }

  /**
   * Whether a value matches a stop value.
   * @param {string} value
   * @returns {Boolean}
   */
  isStopped(value) {
    return this.stops.some((pattern) => matchesRegex(pattern, value));
  }

  /**
   * Breadth-first closure from a seed value over the two-way join.
   * Returns newly reached values in discovery order, skipping stops
   * and values this traversal has already expanded.
   * @param {string} seed
   * @returns {string[]}
   */
  closure(seed) {
    const reached = [];

    if (this.visited.has(seed)) return reached;

    this.visited.add(seed);

    const queue = [seed];

    while (queue.length > 0) {
      const value = queue.shift();

      const nexts = this.neighbors.get(value);

      if (nexts === undefined) continue;

      for (const next of nexts) {
        if (this.visited.has(next) || this.isStopped(next)) continue;

        this.visited.add(next);

        reached.push(next);

        queue.push(next);
      }
    }

    return reached;
  }
}

/**
 * Parse `~` specs out of the query, removing the `~` key.
 * Each spec names a field of the query base collection and optionally
 * carries stop regexes under `.`. A spec without a base value or with
 * a base other than `~` is discarded.
 * @returns {{ traversals: Traversal[], stripped: object }}
 */
export async function extractTraversals(fs, dir, schema, query) {
  const { "~": specs, ...stripped } = query;

  if (specs === undefined) return { traversals: [], stripped };

  const specList = Array.isArray(specs) ? specs : [specs];

  const traversals = [];

  for (const specConcise of specList) {
    // concise form: "actname" is short for { _: "~", "~": "actname" }
    const spec =
      typeof specConcise === "string"
        ? { _: "~", "~": specConcise }
        : specConcise;

    if (spec === null || typeof spec !== "object") continue;

    if (spec._ !== "~") continue;

    // the base value of a `~` record is an exact field name
    const field = spec["~"];

    if (typeof field !== "string" || field === "") continue;

    // stop values are regexes; a malformed regex matches nothing
    const stopSpecs =
      spec["."] === undefined
        ? []
        : Array.isArray(spec["."])
          ? spec["."]
          : [spec["."]];

    const stops = stopSpecs
      .map((stop) =>
        typeof stop === "string"
          ? stop
          : stop !== null && typeof stop === "object"
            ? stop[stop._]
            : undefined,
      )
      .filter((stop) => typeof stop === "string");

    const neighbors = await loadNeighbors(fs, dir, schema, query._, field);

    traversals.push(new Traversal(stops, neighbors));
  }

  return { traversals, stripped };
}

/**
 * Find the tablet that stores the relationship between `base` and
 * `field` and read it into a two-way adjacency map. An unknown
 * relationship or a missing tablet yields an empty map.
 * @returns {Map<string, string[]>}
 */
async function loadNeighbors(fs, dir, schema, base, field) {
  const neighbors = new Map();

  // collection names are interpolated into a filename
  if (!isWellFormedBranch(base) || !isWellFormedBranch(field)) {
    return neighbors;
  }

  const hasTrunk = (branch, trunk) =>
    schema[branch] !== undefined && schema[branch].trunks.includes(trunk);

  let filename;

  if (hasTrunk(field, base)) {
    filename = `${base}-${field}.csv`;
  } else if (hasTrunk(base, field)) {
    filename = `${field}-${base}.csv`;
  } else {
    return neighbors;
  }

  const push = (key, value) => {
    const values = neighbors.get(key);

    if (values === undefined) {
      neighbors.set(key, [value]);
    } else {
      values.push(value);
    }
  };

  for await (const group of keyGroups(fs, dir, filename)) {
    for (const value of group.values) {
      push(group.key, value);

      push(value, group.key);
    }
  }

  return neighbors;
}

File: rs/src/dataset/select/recursion.rs
use crate::dataset::query::groups::key_groups;
use crate::schema::is_well_formed_branch;
use crate::{Branch, Entry, Result, Schema, Trunks};
use regex::Regex;
use std::collections::{HashMap, HashSet, VecDeque};
use std::path::Path;

/// One `~` traversal: a closure over the tablet that stores the
/// relationship between the query base and the named field.
pub struct Traversal {
    /// Stop regexes: a value matching any of these is neither
    /// returned nor expanded.
    stops: Vec<Regex>,
    /// Two-way adjacency over the tablet lines. The direction in
    /// which the tablet stores the relationship does not affect
    /// the result.
    neighbors: HashMap<String, Vec<String>>,
    /// Values this traversal has already expanded.
    visited: HashSet<String>,
}

impl Traversal {
    /// Whether a value matches a stop value.
    pub fn is_stopped(&self, value: &str) -> bool {
        self.stops.iter().any(|re| re.is_match(value))
    }

    /// Breadth-first closure from a seed value over the two-way join.
    /// Returns newly reached values in discovery order, skipping stops
    /// and values this traversal has already expanded.
    pub fn closure(&mut self, seed: &str) -> Vec<String> {
        let mut reached = vec![];

        if !self.visited.insert(seed.to_owned()) {
            return reached;
        }

        let mut queue = VecDeque::from([seed.to_owned()]);

        while let Some(value) = queue.pop_front() {
            let nexts = match self.neighbors.get(&value) {
                Some(ns) => ns.clone(),
                None => continue,
            };

            for next in nexts {
                if self.visited.contains(&next) || self.is_stopped(&next) {
                    continue;
                }

                self.visited.insert(next.clone());

                reached.push(next.clone());

                queue.push_back(next);
            }
        }

        reached
    }
}

/// Parse `~` specs out of the query leaves, removing the `~` leaf.
/// Each spec names a field of the query base collection and optionally
/// carries stop regexes under `.`. A spec without a base value or with
/// a base other than `~` is discarded.
pub fn extract_traversals(
    dir: &Path,
    schema: &Schema,
    query: &mut Entry,
) -> Result<Vec<Traversal>> {
    let specs = match query.leaves.remove("~") {
        None => return Ok(vec![]),
        Some(specs) => specs,
    };

    let mut traversals = vec![];

    for spec in specs {
        if spec.base != "~" {
            continue;
        }

        // the base value of a `~` record is an exact field name
        let field = match &spec.base_value {
            Some(f) if !f.is_empty() => f.clone(),
            _ => continue,
        };

        // stop values are regexes; a malformed regex matches nothing
        let stops: Vec<Regex> = spec
            .leaves
            .get(".")
            .map(|vs| {
                vs.iter()
                    .filter_map(|v| v.base_value.as_deref())
                    .filter_map(|s| Regex::new(s).ok())
                    .collect()
            })
            .unwrap_or_default();

        let neighbors = load_neighbors(dir, schema, &query.base, &field)?;

        traversals.push(Traversal {
            stops,
            neighbors,
            visited: HashSet::new(),
        });
    }

    Ok(traversals)
}

/// Find the tablet that stores the relationship between `base` and
/// `field` and read it into a two-way adjacency map. An unknown
/// relationship or a missing tablet yields an empty map.
fn load_neighbors(
    dir: &Path,
    schema: &Schema,
    base: &str,
    field: &str,
) -> Result<HashMap<String, Vec<String>>> {
    // collection names are interpolated into a filename
    if !is_well_formed_branch(base) || !is_well_formed_branch(field) {
        return Ok(HashMap::new());
    }

    let has_trunk = |branch: &str, trunk: &str| match schema.0.get(branch) {
        Some(Branch {
            trunks: Trunks(ts), ..
        }) => ts.iter().any(|t| t == trunk),
        None => false,
    };

    let filename = if has_trunk(field, base) {
        format!("{}-{}.csv", base, field)
    } else if has_trunk(base, field) {
        format!("{}-{}.csv", field, base)
    } else {
        return Ok(HashMap::new());
    };

    let mut neighbors: HashMap<String, Vec<String>> = HashMap::new();

    for group in key_groups(&dir.join(filename))? {
        for value in group.values {
            neighbors
                .entry(group.key.clone())
                .or_default()
                .push(value.clone());

            neighbors.entry(value).or_default().push(group.key.clone());
        }
    }

    Ok(neighbors)
}
